### PR TITLE
FOFB system identification features: ABI

### PR DESCRIPTION
**Pull request by guilhermerc**
PR for `fofb_sys_id` first production ABI.

### CODE REVIEW

**Comment by guilhermerc** [comment]
@augustofg , @ericonr , I've opened this draft PR so we can discuss `fofb_sys_id` ABI.

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h [hunk 1]
```diff
@@ -39,6 +119,88 @@ positions. The valid range of this register is [0, 255].
...
+
+      /* [0x800]: MEMORY Distortion levels RAM */
+      struct distort_ram {
+        /* [0x0]: REG (rw) Two S16FP0-bit distortion levels, one for each PRBS
+value.
+
+15 - 0: distortion level for PRBS value 0
+31 - 16: distortion level for PRBS value 1
+ */
+        uint32_t levels;
```
Comment (ericonr) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h:175]: Could we take advantage of the smaller registers here to have two `uint16_t` values?

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h [hunk 2]
```diff
@@ -39,6 +119,88 @@ positions. The valid range of this register is [0, 255].
...
+      /* [0x40]: REPEAT Distortion levels registers (one for each channel)
+ */
+      struct ch {
+        /* [0x0]: REG (rw) Two S16FP0-bit distortion levels, one for each PRBS
+value.
+
+15 - 0: distortion level for PRBS value 0
+31 - 16: distortion level for PRBS value 1
+ */
+        uint32_t levels;
```
Comment (ericonr) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h:197]: Same here.

**Review by ericonr — COMMENTED** [review]
There are some ~4000 and 511 words padding that seem to be a bit too big

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h (hunk 2 above)
Comment (guilhermerc) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h:197]: For consistency with `prbs_bpm_pos_distort` block, I kept a single register here too.

Comment (guilhermerc) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h:197]: I actually don't mind breaking the consistency and using 2x16-bit registers for the levels here. What do you think?

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby [hunk 3]
```diff
@@ -28,3 +28,153 @@ memory-map:
...
+                                range: 15-0
+                                description: |
+                                  Distortion level for PRBS value 0 in
+                                  nanometers
+                            - field:
+                                name: level_1
+                                range: 31-16
+                                description: |
+                                  Distortion level for PRBS value 1 in
+                                  nanometers
```
Comment (augustofg) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:130]: Move the memory block to the end of the cheby file to minimize the amount of padding required.

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby [hunk 4]
```diff
@@ -28,3 +28,153 @@ memory-map:
...
+                    count: 12
+                    description: |
+                      Distortion levels registers (one for each channel)
+                    children:
+                      - reg:
+                          name: levels
+                          width: 32
+                          access: rw
+                          description: |
+                            Two S16FP0-bit distortion levels, one for each PRBS
```
Comment (augustofg) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:163]: I find using this `S16FP0` notation confusing here when you can simply write 'Two signed 16 bits...'

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby [hunk 5]
```diff
@@ -28,3 +28,153 @@ memory-map:
...
+
+                            15 - 0: distortion level for PRBS value 0
+                            31 - 16: distortion level for PRBS value 1
+                          children:
+                            - field:
+                                name: level_0
+                                range: 15-0
+                                description: |
+                                  Distortion level for PRBS value 0 in
+                                  nanometers
```
Comment (augustofg) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:174]: Not nanometers, but ADC counts of current of the RTM-LAMP power supply.

**Review by augustofg — CHANGES_REQUESTED** [review]

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h (hunk 2 above)
Comment (ericonr) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h:197]: For the higher level I think I'd rather keep the consistency, then. Thanks!

Comment (guilhermerc) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.h:197]: I did this ^

**Comment by guilhermerc** [comment]
I've addressed all the requested changes.

**Review by ericonr — APPROVED** [review]

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby [hunk 6]
```diff
@@ -7,24 +7,160 @@ memory-map:
...
+                - field:
+                    name: rst
+                    x-hdl:
+                      type: autoclear
+                    range: 0
+                    description: |
+                      Resets PRBS
+
+                      write 0: no effect
+                      write 1: resets PRBS (this bit autoclears)
```
Comment (ericonr) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:60]: This has to be updated: "enables PRBS reset via trigger", and it doesn't autoclear.

Should we document the logical trigger (probably from the SYSID trigger mux?) responsible for this here? Or somewhere else?

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby [hunk 7]
```diff
@@ -7,24 +7,160 @@ memory-map:
...
+                      write 0: distortion disabled
+                      write 1: distortion enabled
+                - field:
+                    name: sp_distort_en
+                    range: 17
+                    description: |
+                      Enables/disables PRBS-based distortion on setpoints
+
+                      write 0: distortion disabled
+                      write 1: distortion enabled
```
Comment (ericonr) [inline hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:98]: Can they be enabled simultaneously? If not, that should be documented here.

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby (hunk 6 above)
Comment (augustofg) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:60]: Not here, the way you connect the trigger is defined in the top level RTL entity.

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby (hunk 7 above)
Comment (augustofg) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:98]: They can, though I don't know how useful it is.

Comment (guilhermerc) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:98]: They can, @ericonr .

File: hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby (hunk 6 above)
Comment (guilhermerc) [reply hdl/modules/fofb_sys_id/cheby/wb_fofb_sys_id_regs.cheby:60]: @ericonr , I'll add a new commit with trigger support.

**Comment by augustofg** [comment]
:+1:

**Pull request merged**